diff --git a/includes/field_ui.inc b/includes/field_ui.inc
index 6a3cf76..2cfddfa 100644
--- a/includes/field_ui.inc
+++ b/includes/field_ui.inc
@@ -455,8 +455,10 @@ function field_group_format_settings_form(&$group, $form, $form_state) {
     'configuration' => array('label' => $group->label, 'settings' => $group->format_settings),
     'group' => $group,
   ));
-
-  return $plugin->settingsForm($form, $form_state);
+  if ($plugin) {
+    return $plugin->settingsForm($form, $form_state);
+  }
+  return $form;
 }
 
 /**
@@ -505,13 +507,15 @@ function field_group_format_settings_summary($group_name, $group) {
     'configuration' => array('label' => $group->label, 'settings' => $group->format_settings),
     'group' => $group,
   ));
+  if($plugin) {
+    $summary = $plugin->settingsSummary();
 
-  $summary = $plugin->settingsSummary();
-
-  return array(
-    '#markup' => '<div class="field-plugin-summary">' . implode('<br />', $summary) . '</div>',
-    '#cell_attributes' => array('class' => array('field-plugin-summary-cell')),
-  );
+    return array(
+      '#markup' => '<div class="field-plugin-summary">' . implode('<br />', $summary) . '</div>',
+      '#cell_attributes' => array('class' => array('field-plugin-summary-cell')),
+    );
+  }
+  return [];
 }
 
 /**
