diff --git a/field_group.module b/field_group.module
index e85c489..15cddd1 100644
--- a/field_group.module
+++ b/field_group.module
@@ -285,7 +285,9 @@ function field_group_field_group_pre_render(&$element, &$group, &$rendering_obje
     'configuration' => array('label' => $group->label, 'settings' => $group->format_settings),
     'group' => $group,
   ));
-  $plugin->preRender($element, $rendering_object);
+  if ($plugin) {
+    $plugin->preRender($element, $rendering_object);
+  }
 
 }
 
